source/sigv4.c

```diff
@@ -793,6 +841,231 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+    {
+        size_t noOfHeaders = 0, buffRemaining = 0, keyLen = 0, valLen = 0, i = 0, trimValueLen = 0;
+        char * pBufLoc;
+        const char * value;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        for( noOfHeaders = 0; noOfHeaders < SIGV4_MAX_HTTP_HEADER_COUNT; noOfHeaders++ )
+        {
+            if( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData != NULL )
+            {
+                sigV4Status = writeCanonicalHeaderToString( noOfHeaders, canonicalRequest );
+
+                if( sigV4Status != SigV4Success )
+                {
+                    break;
+                }
+            }
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t generateCanonicalHeaders( const char * pHeaders,
+                                                   size_t headersLen,
+                                                   CanonicalContext_t * canonicalRequest )
+    {
+        const char * start;
+        const char * end;
+        size_t fieldFlag = 1, noOfHeaders = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( pHeaders != NULL );
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        start = end = pHeaders;
+
+        for( i = 0; i < headersLen; i++ )
+        {
+            if( noOfHeaders == SIGV4_MAX_HTTP_HEADER_COUNT )
+            {
+                return SigV4MaxHeaderPairCountExceeded;
+            }
+
+            /* Extracting each header key and value from the headers string. */
+            if( fieldFlag == 1 )
+            {
+                if( pHeaders[ i ] == ':' )
+                {
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData = start;
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].key.dataLen = ( end - start );
+                    end++;
+                    start = end;
+                    fieldFlag = 0;
+                }
+                else
+                {
+                    end++;
+                }
+            }
+            else
+            {
+                if( pHeaders[ i ] == '\n' )
+                {
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = start;
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( end - start );
+                    end++;
+                    start = end;
+                    fieldFlag = 1;
+                    noOfHeaders++;
+                }
+                else
+                {
+                    end++;
+                }
+            }
+        }
+
+        /* Sorting headers based on keys. */
+        qsort( canonicalRequest->pHeadersLoc, 5, sizeof( SigV4KeyValuePair_t ), cmpKeyValue );
```

Comment:
The number of keys here are fixed

Reply:
The function name `cmpKeyValue` is a little misleading as the headers only need to be sorted by field
```suggestion
        qsort( canonicalRequest->pHeadersLoc, 5, sizeof( SigV4KeyValuePair_t ), cmpField );
```